### DISCUSSION TITLE
[HELP NEEDED] How Do I Deploy My C++ Project with IMGUI-Bundle Using Emscripten for the Web?

### DISCUSSION THREAD

**Discussion by SPyofgame200** [pthom/imgui_bundle#274] [General]
## Dear...

Maintainers/Support Team/Community/Experts

## Summary:

I am working on a C++ project that uses the IMGUI-Bundle for creating a visualization application with a suite of tools for end-users. This is a free educational application that help the teachers and their students for demonstrating the Algorithms and Data Structure in Graph Theory. I need to deploy this project to a web platform using Emscripten, but as a first-time user of Emscripten, I am encountering multiple challenges that hinder my progress. While I have followed several online tutorials, they seem to lack specific steps necessary to complete a successful deployment. While some other youtubers teach a way to make C++ games to be deployed on website, most of them using Raylib, HelloImGUI or something simple, I want to deploy my application that using IMGUI-Bundle for more features. Additionally, the official documentation for IMGUI-Bundle provides limited guidance on using Emscripten for deployment, leaving gaps in the workflow I need to follow.

## Context:

My project requires deploying a C++ visualization tool on the web to ensure accessibility and ease of use for a broader audience. The IMGUI-Bundle library has been effective for local development and provides the GUI capabilities I need. However, transitioning this environment from a desktop application to a web-based one, particularly using Emscripten, is an unfamiliar process. This transition is essential for the project’s goal to offer a flexible, accessible, and platform-independent toolset.

## Challenges:

1. Since this is my first exposure to Emscripten, I am finding it challenging to understand its core concepts and setup procedures. Specific steps involved in configuring Emscripten to build and deploy an IMGUI-Bundle-based C++ project to a web platform are unclear.
2. Most online tutorials and video guides are either outdated or omit crucial details, leading to build failures and incomplete deployments. The IMGUI-Bundle documentation also lacks specific instructions on deploying projects using Emscripten, which makes integrating these two tools even more difficult.
3. I have encountered various build errors that I am not equipped to troubleshoot effectively, given my limited experience with Emscripten. These include missing dependencies, linker errors, and configuration mismatches between IMGUI-Bundle and the Emscripten environment.

## I have tried...

1. Followed multiple video tutorials and blog posts about deploying C++ applications using Emscripten, noting common steps but finding that each left out critical details, resulting in deployment failures.
2. Attempted to adapt the IMGUI-Bundle’s setup instructions for desktop builds to a web environment, but the lack of documentation specific to Emscripten integration led to additional issues.
3. Examined the official Emscripten and IMGUI-Bundle documentation to piece together an integration workflow, but found it challenging to interpret without prior Emscripten experience.
4. I have been working on my project for 3 weeks, trying to deploy it to a website with no real progress. I keep getting set back whenever I encounter issues that the YouTube tutorials don’t cover or the documentation doesn’t address.
5. Trying to deploy a project with IMGUI thne replace EACH INDIVIDUAL file or folder with IMGUI-Bundle, but still failed after repeating the steps again.
6.  Search for similar disscussions and solutions (like ideas, show and tells, ...) in IMGUI-BUNDLE, IMGUI, discords or some possible deployed websites with IMGUI.

## I am seeking guidance on...

1. A clear, step-by-step process to set up, build, and deploy a C++ project using IMGUI-Bundle with Emscripten, including any necessary configurations, dependencies, and command-line instructions. Maybe with a video included.
2. Specific guidance on adapting IMGUI-Bundle’s setup to work seamlessly within an Emscripten environment, particularly for web-based deployment. Life what might be shown on C++ Application but not on the web site, what are the imcompatibilities of them, and what are the pros and cons of using Empscripten.
3. Recommendations for troubleshooting common build issues encountered when using Emscripten with C++ projects, especially for UI-focused applications like this one.  If possible, I would also appreciate alternative approaches to avoid hidden assumptions (e.g., prerequisites or setup steps that may not be immediately obvious).
4.  Some possible updates for the documentation to cover this area ASAP.
5. Suggestions on how I can accomplish these tasks for free if possible (including website hosting, etc.), as I currently don’t have a budget.

## Conclusions

Your insights on these challenges would be invaluable for helping me successfully deploy this application and make more users try the demos before cloning my application.

Thank you very much for your time and assistance.

Best ragards,
Võ Hoàng Anh (SPyofgame)

**Comment by pthom** [comment]
Hello,

You may want to have a look at the [RPN calculator demo](https://pthom.github.io/hello_imgui/book/intro.html#rpn-calculator) which explains multiplatform deployment (including emscripten)

[Online demo](https://traineq.org/HelloImGui/rpn/rpn_calculator.html) - [Source](https://github.com/pthom/rpn_calculator/) - [Video tutorial](https://www.youtube.com/watch?v=dArP4lBnOr8)

This tutorial is for hello _imgui but the principle is almost identical with imgui_bundle, except that in CMake you call `imgui_bundle_add_app` instead of `hello_imgui_add_app`.

**Reply by digitalsignalperson** [reply]
@pthom do you have any docs for taking something like that RPN calculator written in C++ and adding it to the imgui_bundle python bidings to work with python projects natively or with pyodide?

I find it easiest to write imgui code in C++, but also want to be able to make it easy for users that only know python to easily plug in my custom widgets or tools. Just wanted to ask in case you have a good example of this somewhere. Thanks for the great work!

**Reply by pthom** [reply]
@digitalsignalperson : you might want to read https://pthom.github.io/imgui_bundle/devel_docs/bindings.html

**Comment by digitalsignalperson** [comment]
Perfect exactly what I was looking for. Thanks
